Fix flaky unit test

### Problem encountered

The `TestLoadingDataFetchedAsynchronously` test (datanode/networkhistory) fails fairly regularly, and should be made more stable. The test can fail both in the regular unit test run, and the run with race detection.

### Observed behaviour

Looking at the logs, we can see the test fails because of a deadlock on the DB side:

```
2024-01-04T19:23:13.037Z] 2024-01-04T19:21:02.404Z	INFO	revert-schema-to-version-0	v3@v3.6.1/migration.go:46	OK   0062_teams_stats.sql
[2024-01-04T19:23:13.037Z]     service_test.go:431: 
[2024-01-04T19:23:13.037Z]         	Error Trace:	/jenkins/workspace/vega_PR-10315/vega/datanode/networkhistory/service_test.go:431
[2024-01-04T19:23:13.037Z]         	Error:      	Received unexpected error:
[2024-01-04T19:23:13.037Z]         	            	failed to load snapshot data:failed to revert scheam to version zero: failed to goose down the schema to version 0: ERROR 0061_additional_candle_intervals.sql: failed to run SQL migration: failed to execute SQL query "SELECT remove_continuous_aggregate_policy('trades_candle_4_hours', true);\n": ERROR: deadlock detected (SQLSTATE 40P01)
[2024-01-04T19:23:13.037Z]         	Test:       	TestLoadingDataFetchedAsynchronously
[2024-01-04T19:23:13.037Z] --- FAIL: TestLoadingDataFetchedAsynchronously (4.91s)
```

Looking at the specific error message, and the migration itself (removing a continuous aggregate policy), the likely cause is a pgsql jobworker is using the underlying materialized view while we're trying to delete it, causing a deadlock. During the down step, we might be able to mitigate the issue by disabling the scheduler, although this is speculation at this point.

### Expected behaviour

The test passing.

### Steps to reproduce

```bash
run the datanode/network unit tests a number of times with `-count=1` to disable caching, the failure is intermittent.
```


### Software version

develop

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -117,6 +117,7 @@
 - [10354](https://github.com/vegaprotocol/vega/issues/10354) - Renumbered SQL migration scripts 0055-0067 due to 0068 being released as part of a patch without renumbering.
 - [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
 - [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix for GraphQL batch proposals support.
+- [10326](https://github.com/vegaprotocol/vega/issues/10326) - Fix intermittent test failure.
 
 ## 0.73.0
 

--- a/core/validators/witness_test.go
+++ b/core/validators/witness_test.go
@@ -18,9 +18,11 @@ package validators_test
 import (
 	"context"
 	"encoding/hex"
+	"sync"
 	"testing"
 	"time"
 
+	"code.vegaprotocol.io/vega/core/txn"
 	"code.vegaprotocol.io/vega/core/validators"
 	"code.vegaprotocol.io/vega/core/validators/mocks"
 	"code.vegaprotocol.io/vega/libs/crypto"
@@ -28,7 +30,9 @@ import (
 	"code.vegaprotocol.io/vega/logging"
 	commandspb "code.vegaprotocol.io/vega/protos/vega/commands/v1"
 
+	"github.com/cenkalti/backoff"
 	"github.com/golang/mock/gomock"
+	"github.com/golang/protobuf/proto"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -245,21 +249,30 @@ func TestVoteMajorityCalculation(t *testing.T) {
 	// first wait once for the asset to be validated
 	<-ch
 
+	wg := sync.WaitGroup{}
+	wg.Add(2)
 	// first on chain time update, we send our own vote
-	erc.cmd.EXPECT().Command(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
+	erc.cmd.EXPECT().Command(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Do(func(_ context.Context, _ txn.Command, _ proto.Message, _ func(string, error), _ *backoff.ExponentialBackOff) {
+		wg.Done()
+	})
 	newNow := erc.startTime.Add(1 * time.Second)
 	erc.OnTick(context.Background(), newNow)
 
 	// then we propagate our own vote
 	pubKey := newPublicKey(selfPubKey)
-	erc.top.EXPECT().IsValidatorVegaPubKey(pubKey.Hex()).Times(1).Return(true)
+	erc.top.EXPECT().IsValidatorVegaPubKey(pubKey.Hex()).Times(1).Return(true).Do(func(_ string) {
+		wg.Done()
+	})
 	err = erc.AddNodeCheck(context.Background(), &commandspb.NodeVote{Reference: res.id}, pubKey)
 	assert.NoError(t, err)
 
 	for i := 0; i < 3; i++ {
 		// second vote from another validator
 		othPubKey := newPublicKey(crypto.RandomHash())
-		erc.top.EXPECT().IsValidatorVegaPubKey(othPubKey.Hex()).Times(1).Return(true)
+		wg.Add(1)
+		erc.top.EXPECT().IsValidatorVegaPubKey(othPubKey.Hex()).Times(1).Return(true).Do(func(_ string) {
+			wg.Done()
+		})
 		err = erc.AddNodeCheck(context.Background(), &commandspb.NodeVote{Reference: res.id}, othPubKey)
 		assert.NoError(t, err)
 	}
@@ -272,6 +285,7 @@ func TestVoteMajorityCalculation(t *testing.T) {
 
 	// block to wait for the result
 	<-ch
+	wg.Wait()
 }
 
 func testOnTick(t *testing.T) {

--- a/datanode/networkhistory/service.go
+++ b/datanode/networkhistory/service.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"io"
 	"sort"
+	"strings"
 	"sync"
 	"time"
 
@@ -261,6 +262,9 @@ func (d *Service) LoadNetworkHistoryIntoDatanode(ctx context.Context, chunk segm
 func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log snapshot.LoadLog, chunk segment.ContiguousHistory[segment.Full],
 	connConfig sqlstore.ConnectionConfig, withIndexesAndOrderTriggers, verbose bool,
 ) (snapshot.LoadResult, error) {
+	maxRetries := 2 // max 2 retries
+	// the deadlock error that should trigger a retry
+	status := "deadlock detected (SQLSTATE 40P01)"
 	datanodeBlockSpan, err := sqlstore.GetDatanodeBlockSpan(ctx, d.connPool)
 	if err != nil {
 		return snapshot.LoadResult{}, fmt.Errorf("failed to get data node block span: %w", err)
@@ -274,14 +278,29 @@ func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log
 
 	chunks := chunk.Slice(datanodeBlockSpan.ToHeight+1, chunk.HeightTo)
 	loadResult, err := d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
-	if err != nil {
+	if err == nil {
+		log.Info("loaded all available data into datanode", logging.String("result", fmt.Sprintf("%+v", loadResult)),
+			logging.Duration("time taken", time.Since(start)))
+
+		return loadResult, err
+	}
+	if !strings.Contains(err.Error(), status) {
+		// some error other than 40P01 encountered
 		return snapshot.LoadResult{}, fmt.Errorf("failed to load snapshot data:%w", err)
 	}
-
-	log.Info("loaded all available data into datanode", logging.String("result", fmt.Sprintf("%+v", loadResult)),
-		logging.Duration("time taken", time.Since(start)))
-
-	return loadResult, err
+	for retries := 0; retries < maxRetries; retries++ {
+		loadResult, err = d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
+		if err == nil {
+			log.Info("loaded all available data into datanode WITH RETRIES",
+				logging.String("result", fmt.Sprintf("%+v", loadResult)),
+				logging.Duration("time taken", time.Since(start)),
+				logging.Int("retry-count", retries+1),
+			)
+			return loadResult, err
+		}
+	}
+	// retries still ended up failing
+	return snapshot.LoadResult{}, fmt.Errorf("failed to load snapshot data:%w", err)
 }
 
 func (d *Service) GetMostRecentHistorySegmentFromBootstrapPeers(ctx context.Context,

--- a/datanode/networkhistory/service.go
+++ b/datanode/networkhistory/service.go
@@ -262,7 +262,7 @@ func (d *Service) LoadNetworkHistoryIntoDatanode(ctx context.Context, chunk segm
 func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log snapshot.LoadLog, chunk segment.ContiguousHistory[segment.Full],
 	connConfig sqlstore.ConnectionConfig, withIndexesAndOrderTriggers, verbose bool,
 ) (snapshot.LoadResult, error) {
-	maxRetries := 3 // max 2 retries
+	maxRetries := 3
 	// the deadlock error that should trigger a retry
 	status := "deadlock detected (SQLSTATE 40P01)"
 	datanodeBlockSpan, err := sqlstore.GetDatanodeBlockSpan(ctx, d.connPool)

--- a/datanode/networkhistory/service.go
+++ b/datanode/networkhistory/service.go
@@ -261,7 +261,7 @@ func (d *Service) LoadNetworkHistoryIntoDatanode(ctx context.Context, chunk segm
 
 func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log snapshot.LoadLog, chunk segment.ContiguousHistory[segment.Full],
 	connConfig sqlstore.ConnectionConfig, withIndexesAndOrderTriggers, verbose bool,
-) (loadResult snapshot.LoadResult, err error) {
+) (snapshot.LoadResult, error) {
 	maxRetries := 3 // max 2 retries
 	// the deadlock error that should trigger a retry
 	status := "deadlock detected (SQLSTATE 40P01)"
@@ -276,24 +276,27 @@ func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log
 
 	start := time.Now()
 
+	var rErr error // return error
 	chunks := chunk.Slice(datanodeBlockSpan.ToHeight+1, chunk.HeightTo)
 	for retries := 0; retries < maxRetries; retries++ {
-		loadResult, err = d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
+		loadResult, err := d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
 		if err == nil {
 			log.Info("loaded all available data into datanode",
 				logging.String("result", fmt.Sprintf("%+v", loadResult)),
 				logging.Duration("time taken", time.Since(start)),
 				logging.Int("retry-count", retries),
 			)
-			return loadResult, err
+			return loadResult, nil
 		}
+		// keep track of the last error
+		rErr = err
 		if !strings.Contains(err.Error(), status) {
 			// some error other than 40P01 encountered
 			break
 		}
 	}
 	// retries still ended up failing
-	return snapshot.LoadResult{}, fmt.Errorf("failed to load snapshot data:%w", err)
+	return snapshot.LoadResult{}, fmt.Errorf("failed to load snapshot data:%w", rErr)
 }
 
 func (d *Service) GetMostRecentHistorySegmentFromBootstrapPeers(ctx context.Context,

--- a/datanode/networkhistory/service.go
+++ b/datanode/networkhistory/service.go
@@ -261,8 +261,8 @@ func (d *Service) LoadNetworkHistoryIntoDatanode(ctx context.Context, chunk segm
 
 func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log snapshot.LoadLog, chunk segment.ContiguousHistory[segment.Full],
 	connConfig sqlstore.ConnectionConfig, withIndexesAndOrderTriggers, verbose bool,
-) (snapshot.LoadResult, error) {
-	maxRetries := 2 // max 2 retries
+) (loadResult snapshot.LoadResult, err error) {
+	maxRetries := 3 // max 2 retries
 	// the deadlock error that should trigger a retry
 	status := "deadlock detected (SQLSTATE 40P01)"
 	datanodeBlockSpan, err := sqlstore.GetDatanodeBlockSpan(ctx, d.connPool)
@@ -277,26 +277,19 @@ func (d *Service) LoadNetworkHistoryIntoDatanodeWithLog(ctx context.Context, log
 	start := time.Now()
 
 	chunks := chunk.Slice(datanodeBlockSpan.ToHeight+1, chunk.HeightTo)
-	loadResult, err := d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
-	if err == nil {
-		log.Info("loaded all available data into datanode", logging.String("result", fmt.Sprintf("%+v", loadResult)),
-			logging.Duration("time taken", time.Since(start)))
-
-		return loadResult, err
-	}
-	if !strings.Contains(err.Error(), status) {
-		// some error other than 40P01 encountered
-		return snapshot.LoadResult{}, fmt.Errorf("failed to load snapshot data:%w", err)
-	}
 	for retries := 0; retries < maxRetries; retries++ {
 		loadResult, err = d.snapshotService.LoadSnapshotData(ctx, log, chunks, connConfig, withIndexesAndOrderTriggers, verbose)
 		if err == nil {
-			log.Info("loaded all available data into datanode WITH RETRIES",
+			log.Info("loaded all available data into datanode",
 				logging.String("result", fmt.Sprintf("%+v", loadResult)),
 				logging.Duration("time taken", time.Since(start)),
-				logging.Int("retry-count", retries+1),
+				logging.Int("retry-count", retries),
 			)
 			return loadResult, err
+		}
+		if !strings.Contains(err.Error(), status) {
+			// some error other than 40P01 encountered
+			break
 		}
 	}
 	// retries still ended up failing

